mmgen/models/architectures/sngan_proj/generator_discriminator.py

```diff
@@ -0,0 +1,436 @@
+
+        self.conv_blocks = nn.ModuleList()
+        for idx in range(len(downsample_cfg)):
+            factor_input = 1 if idx == 0 else self.channel_factor_list[idx - 1]
+            factor_output = self.channel_factor_list[idx]
+
+            # get block-specific config
+            block_cfg_ = deepcopy(self.blocks_cfg)
+            block_cfg_['downsample'] = self.downsample_list[idx]
+            block_cfg_['in_channels'] = factor_input * base_channels
+            block_cfg_['out_channels'] = factor_output * base_channels
+            self.conv_blocks.append(build_module(block_cfg_))
+
+        # TODO: add an argument to select bias
+        # self.decision = nn.Linear(
+        #     factor_output * base_channels, 1, bias=False)
+        self.decision = nn.Linear(factor_output * base_channels, 1, bias=True)
+        if with_spectral_norm:
+            self.decision = spectral_norm(self.decision)
+
+        self.num_classes = num_classes
+
+        # In this case, discriminator is designed for conditional synthesis.
+        if num_classes > 0:
+            self.proj_y = nn.Embedding(num_classes,
+                                       factor_output * base_channels)
+            if with_spectral_norm:
+                self.proj_y = spectral_norm(self.proj_y)
+
+        self.activate = build_activation_layer(act_cfg)
+        self.init_weights()
+
+    def forward(self, x, label=None):
+        """Forward function. If ``self.num_classes`` is larger than 0, label
+        projection would be used.
+
+        Args:
+            x (torch.Tensor): Fake or real image tensor.
+            label (torch.Tensor, options): Label correspond to the input image.
+                Noted that, if ``self.num_classed`` is larger than 0,
+                ``label`` should not be None.  Default to None.
+
+        Returns:
+            torch.Tensor: Prediction for the reality of the input image.
+        """
+        h = self.from_rgb(x)
+        for conv_block in self.conv_blocks:
+            h = conv_block(h)
+        h = self.activate(h)
+        h = torch.sum(h, dim=[2, 3])
+        out = self.decision(h)
+
+        if self.num_classes > 0:
+            w_y = self.proj_y(label)
+            out = out + torch.sum(w_y * h, dim=1, keepdim=True)
+        return out.view(out.size(0), -1)
+
+    def init_weights(self, pretrained=None, strict=True):
+        """Init weights for models.
+
+        We just use the initialization method proposed in the original paper.
+
+        Args:
+            pretrained (str, optional): Path for pretrained weights. If given
+                None, pretrained weights will not be loaded. Defaults to None.
+        """
+        if isinstance(pretrained, str):
+            logger = get_root_logger()
+            load_checkpoint(self, pretrained, strict=strict, logger=logger)
+        elif pretrained is None:
+            # TODO: support multi init method
+            nn.init.orthogonal_(self.decision.weight)
```

Comment:
Here, you should note that ``ResBlock`` will not be initialized in the generator. Thus, you require that the block itself must contain correct initialization. This behavior is not safe and users (we) will always forget to check the initialization of the intermia blcks.